neuro_extras/main.py

```diff
@@ -221,6 +252,182 @@ def image() -> None:
+                    extract=extract,
+                )
+            while job.status == neuro_api.JobStatus.PENDING:
+                job = await client.jobs.status(job.id)
+                await asyncio.sleep(1.0)
+            async for chunk in client.jobs.monitor(job.id):
+                if not chunk:
+                    break
+                click.echo(chunk.decode(errors="ignore"), nl=False)
+            job = await client.jobs.status(job.id)
+            if job.status == neuro_api.JobStatus.FAILED:
+                logger.error("The copy job has failed due to:")
+                logger.error(f"  Reason: {job.history.reason}")
+                logger.error(f"  Description: {job.history.description}")
+                exit_code = job.history.exit_code
+                if exit_code is None:
+                    exit_code = EX_PLATFORMERROR
+                sys.exit(exit_code)
+            else:
+                logger.info("Successfully copied data")
+    else:
+        TEMP_UNPACK_DIR.mkdir(parents=True, exist_ok=True)
+        tmp_dir_name = tempfile.mkdtemp(dir=str(TEMP_UNPACK_DIR))
+        tmp_dst_url = URL.build(path=(tmp_dir_name + "/"))
+
+        await _nonstorage_cp(source_url, tmp_dst_url)
+        source_url = tmp_dst_url
+        # for clarity
+        source_url_type = UrlType.LOCAL
+
+        # at this point source is always local
+        if extract:
+            # extract to tmp dir
+            file = Path(source_url.path)
+            if file.is_dir():
+                file = list(file.glob("*"))[0]
+            suffixes = file.suffixes
+            if suffixes[-2:] == [".tar", ".gz"] or suffixes[-1] == ".tgz":
+                command = "tar"
+                args = ["zxvf", str(file), "-C", str(tmp_dst_url)]
+            elif suffixes[-2:] == [".tar", ".bz2"] or suffixes[-1] in (".tbz2", ".tbz"):
+                command = "tar"
+                args = ["jxvf", str(file), "-C", str(tmp_dst_url)]
+            elif suffixes[-1] == ".tar":
+                command = "tar"
+                args = ["xvf", str(file), "-C", str(tmp_dst_url)]
+            elif suffixes[-1] == ".gz":
+                command = "gunzip"
+                args = [str(file), str(tmp_dst_url) + file.name[:-3]]
+            elif suffixes[-1] == ".zip":
+                command = "unzip"
+                args = [str(file), "-d", str(tmp_dst_url)]
+            else:
+                raise ValueError(f"Don't know how to extract file {file.name}")
+
+            click.echo(f"Running {command} {' '.join(args)}")
+            subprocess = await asyncio.create_subprocess_exec(command, *args)
+            returncode = await subprocess.wait()
+            if returncode != 0:
+                raise click.ClickException(f"Extraction failed: {subprocess.stderr}")
+            else:
+                if file.exists():
+                    # gunzip removes src after extraction, while tar - not
+                    file.unlink()
```

Comment:
later: better to wrap cleanups with try-catch because its failure doesn't mean that the overall command failed. The cleanup errors should be printed out and suppressed, I think.